crates/viewer/re_blueprint_tree/src/data.rs

```diff
@@ -0,0 +1,515 @@
+                        None
+                    }
+                }
+            })
+            .collect_vec();
+
+        // everything was filtered out
+        if filter_matcher.is_active() && children.is_empty() {
+            return None;
+        }
+
+        Some(Self {
+            id: container_blueprint.id,
+            name: container_blueprint.display_name_or_default(),
+            kind: container_blueprint.container_kind,
+            visible: container_blueprint.visible,
+            default_open: true,
+            children,
+        })
+    }
+}
+
+// ---
+
+/// Data related to a single view and its content.
+#[derive(Debug)]
+#[cfg_attr(feature = "testing", derive(serde::Serialize, serde::Deserialize))]
+pub struct ViewData {
+    pub id: ViewId,
+
+    pub class_identifier: ViewClassIdentifier,
+
+    pub name: ContentsName,
+    pub visible: bool,
+    pub default_open: bool,
+
+    /// The origin tree contains the data results contained in the subtree defined by the view
+    /// origin. They are presented first in the blueprint tree.
+    pub origin_tree: Option<DataResultData>,
+
+    /// Projection trees are the various trees contained data results which are outside the view
+    /// origin subtrees. They are presented after a "Projections:" label in the blueprint tree.
+    ///
+    /// These trees may be super-trees of the view origin trees. In that case, the view origin
+    /// subtree is represented by a stub item, see [`DataResultKind::OriginProjectionPlaceholder`].
+    pub projection_trees: Vec<DataResultData>,
+}
+
+impl ViewData {
+    fn from_blueprint_and_filter(
+        ctx: &ViewerContext<'_>,
+        view_blueprint: &ViewBlueprint,
+        filter_matcher: &FilterMatcher,
+    ) -> Option<Self> {
+        re_tracing::profile_function!();
+
+        let query_result = ctx.lookup_query_result(view_blueprint.id);
+        let result_tree = &query_result.tree;
+
+        //
+        // In origin tree data results
+        //
+
+        let mut hierarchy = Vec::with_capacity(10);
+        let origin_tree = DataResultData::from_data_result_and_filter(
+            ctx,
+            view_blueprint,
+            query_result,
+            &DataResultNodeOrPath::from_path_lookup(result_tree, &view_blueprint.space_origin),
+            false,
+            &mut hierarchy,
+            filter_matcher,
+            false,
+        );
+
+        if cfg!(debug_assertions) {
+            debug_assert!(hierarchy.is_empty());
+        } else {
+            hierarchy.clear();
+        }
+
+        //
+        // Projection data results
+        //
+
+        let mut projections = Vec::new();
+        result_tree.visit(&mut |node| {
+            if node
+                .data_result
+                .entity_path
+                .starts_with(&view_blueprint.space_origin)
+            {
+                // If it's under the origin, we're not interested, stop recursing.
+                false
```

Comment:
Reading this code it's not immediately obvious what returning `false` from `visit` means. It's not a huge issue, but it could probably be made easier to read by using something like [`std::ops::ControlFlow`](https://doc.rust-lang.org/std/ops/enum.ControlFlow.html)

Reply:
Good call! This existing `visit` method is used in a bunch of places so I won't change it here, but I'll keep that in mind and might do a separate cleanup PR.